### DISCUSSION TITLE
Custom markdown elements for using in .md pages

### DISCUSSION THREAD

**Discussion by A3Bagged** [squidfunk/mkdocs-material#7775] [Q&A]
Hello, sorry if this has been implemented already but i couldn't find a fix. I did find the following:
[#514](https://github.com/markedjs/marked/issues/514) and [#7451](https://github.com/squidfunk/mkdocs-material/discussions/7451)
<hr>

Preview:
![image](https://github.com/user-attachments/assets/8b6af416-c204-43fb-a2ba-42cfe1a5986d)

<hr>

Example of a website im trying to copy a function from: [Thulite](https://docs.thulite.io/guides/email-obfuscation/#setup)

Here he uses a ordered list for setup steps, but the markdown is totally different and im trying to implement something like that.
I've tried <steps></steps> and that just throws `<p><steps></steps></p>` above everything else.
It only worked when there was text directly next to `<steps>` without breaks.

Example markdown: [Click here](https://github.com/thuliteio/docs/blob/main/src/content/docs/guides/email-obfuscation.mdx?plain=1)
```
<Steps>

1. Add a default (fallback) email address to `config/_default/params.toml`:

    ```toml title="params.toml"
    # defaultEmail
    defaultEmail = "email@example.com"
    ```

2. Add the following CSS to `assets/scss/common/_custom.scss`:

    ```scss title="_custom.scss"
    span.email b {
      display: none;
    }
    ```

3. Create shortcode file `layouts/shortcodes/email.html` with the following content:

    ```html title="email.html"
    {{- /* Set defaults and get args. */}}
    {{- $address := index .Params 0 | default site.Params.defaultEmail }}

    {{- /* Get parts. */}}
    {{- $addressParts := split $address "@" }}
    {{- $userName := (index $addressParts 0) }}
    {{- $rootDomain := (index $addressParts 1) }}
    {{- $rootDomainParts := split $rootDomain "." }}
    {{- $domainName := (index $rootDomainParts 0) }}
    {{- $topLevelDomain := (index $rootDomainParts 1) }}

    {{- /* Render. */}}
    <span class="email">
      {{- printf "%s@%s<b>.%s</b>.%s" $userName $domainName $domainName $topLevelDomain | safeHTML -}}
    </span>
    ```

    The shortcode gets the email address you provided — using the default email address if you didn’t specify one. Next, it splits the email address in parts — `userName`, `domainName`, and `topLevelDomain` — and renders the HTML.

</Steps>
```

Result: 
![image](https://github.com/user-attachments/assets/15a50dca-567c-4f15-b122-9ff2f3c64a32)
This doesnt render a `<steps>` HTML Block at all but rather creates a `<ol>` list with a role and class applied.

Is this possible with Material? or has it been implemented through a markdown extension? (sorry if it is)

I tried adding:
```
<script type="text/javascript">
    document.registerElement('steps');
    customElements.define("steps", steps);
  </script>
```
But that is for custom HTML blocks that _Thulite_ doesn't use and currently is not being procecced well in the .md file
unless its a small text like you would use `<strong>` and not for paragraphs. ie.
```
<Steps> This here works

But this doesnt
</steps>
```
Results in:
![image](https://github.com/user-attachments/assets/f572c416-fe01-4d90-aaaf-6b043d9060d5)
![image](https://github.com/user-attachments/assets/c0958472-6db1-42dd-bcd3-6030755b641d)

_i've added a red background in css by: `steps, .steps { background:red;}` to debug_

<hr>

I do know that i can change `<step>` to `<ol class="steps">` and style it in CSS but the page writing & markdown won't be used by me, i just do the coding. So using `<steps>` is easier for the user.

Sincerely,  Chris

**Comment by kamilkrzyskow** [comment]
Hello @A3Bagged,
this kind of custom behaviour needs to be implemented as a separate Markdown extension / plugin. The example site you've provided was made with Astro a JavaScript-based site generator which supports JSX / MDX. 
- https://github.com/squidfunk/mkdocs-material/discussions/3621

That's a completely different ecosystem. Additionally, Python Markdown is quite strict when it comes to formatting, and a new empty line typically means there is a new paragraph `<p>`.

Later in https://github.com/squidfunk/mkdocs-material/discussions/7451 squidfunk noted there is hope in the future that Material would support such kind of thing, but it does not currently.

I'm unaware of any Markdown extensions which do the exact thing you want 🤔 You can check the [catalog](https://github.com/mkdocs/catalog). Perhaps you could hack the implementation of the Neoteroi [timeline](https://www.neoteroi.dev/mkdocs-plugins/timeline/) extension.

Apart of an extension you could also try a plugin / [hook](https://www.mkdocs.org/user-guide/configuration/#hooks) which would modify the Makrdown and adjust the HTML formatting to match the strictness of the Python Markdown implementation.

I've done so in the [latest_blog_posts](https://github.com/nypesap/mkdocs-nype/tree/main/mkdocs_nype/plugins/latest_blog_posts) plugin of the [mkdocs-nype](https://github.com/nypesap/mkdocs-nype) theme. 
The code replaces a special token in the Markdown (could be `<steps>`) and then properly formats it for further Markdown processing with very strict Markdown/HTML snippets. Please don't mind the bit of spaghetti code as it survived multiple changes of specification / of my approach 😆 The result can be seen on the [Nype brand site](https://npe.cm/) - [Markdown source](https://github.com/nypesap/nypesap.github.io/blob/7ac27c478cd488372b2a31927da8a8234f93d559/docs/index.md?plain=1#L26).

The plugin / hook approach is simpler to make imo, however it has to scan the Markdown content from scratch, so the "upper" performance is lower than that of an extension, which would work on tokens (I think 😅). iirc the plugin I mentioned only works on the homepage, so I never needed to worry about that.

**Reply by kamilkrzyskow** [reply]
You could also try the new blocks:
- https://facelessuser.github.io/pymdown-extensions/extensions/blocks/ 

I haven't used them myself, but it looks like it supports setting the `tag` name and supports Markdown inside + nesting :thinking:
So I guess it could be used in this sort of scenario.

**Reply by A3Bagged** [reply]
> You could also try the new blocks:
> 
> - https://facelessuser.github.io/pymdown-extensions/extensions/blocks/ 
> 
> 
> 
> I haven't used them myself, but it looks like it supports setting the `tag` name and supports Markdown inside + nesting :thinking:
> 
> So I guess it could be used in this sort of scenario.

Awesome thanks! I'm researching and i think im coming closer. Im new to mkdocs and markdown so knowing what to look for is overwhelming 🤣 i didnfind an [article](https://alberand.com/markdown-custom-tags.html) on custom markdown tags. Does this work on mkdocs? I think as long as the .py file is in the site directory it should work? (Since i think material uses also custom markdown-extensions right?)

Is there a specific directory that you need to follow? (For example material has plugins etc in 'material' folder instead of root)

- chris

**Reply by kamilkrzyskow** [reply]
Normal Python import rules apply here. You can use the `PYTHONPATH` environmental variable to set it to ".", so it will always include the `<current directory>` you're currently running `python` / `mkdocs serve` from.

Then you can create a module, so basically a directory `modulename` with an empty `__init__.py` file inside it. You can add other files beside that init file like `steps.py` and later use `modulename.steps` in your `markdown_extensions`

The docs for the Insiders projects plugin have a brief mention of that approach as well:
- https://squidfunk.github.io/mkdocs-material/plugins/projects/?h=pythonpath#config.projects_config_transform

---

As for the `mkdocs-material` theme and `material` directory, it's all configured during `pip install mkdocs-material` with the [pyproject.toml](https://github.com/squidfunk/mkdocs-material/blob/master/pyproject.toml) file

**Reply by A3Bagged** [reply]
Thanks for the detailed info 😄 
I didnt use --pip install i think but pulled the repo using github desktop. Now i just copy to a new folder inside "github" in documente and rename the old to 'backup' before doing massive changes to the theme so that i can revert. Hopefully all goes well when using build once completed

**Reply by kamilkrzyskow** [reply]
After pulling the repository from GitHub, pip install is still required, but you install it from the directory, not the pypi package.

**Comment by hellt** [comment]
Steps are super useful in the technical docset. A lot of documentation works in the

1. Procedure
2. Outcome

paradigm. And having a visual cue for the procedural steps improves readability and clarity by a lot. I am super interested to see if @A3Bagged you will manage to get some prototype with Blocks API.

I wonder if @facelessuser had some ideas about this as well, or heard similar requests.

**Reply by facelessuser** [reply]
See https://github.com/squidfunk/mkdocs-material/issues/7760#issuecomment-2574390090

**Reply by hellt** [reply]
oh, damn, this looks great already. Thanks a lot @facelessuser, I think it is long overdue for me to start sponsoring your work. Fixed it now.

**Reply by hellt** [reply]
@facelessuser is there an example of taking `/// html | div.steps`-like block converted to a named block like `/// steps`?

The approach you shared works nice, but the challenge is with using `ol` inside the blocks and thus indenting the content.

**Reply by facelessuser** [reply]
> The approach you shared works nice, but the challenge is with using ol inside the blocks and thus indenting the content.

I didn't understand this statement. Can you illustrate what the challenge is?

**Reply by hellt** [reply]
Sure. Here is how the current solution outlined in https://github.com/squidfunk/mkdocs-material/issues/7760#issuecomment-2574390090 works:

```
/// html | div.steps

1. Step 1

    Some text that must be indented with 4 spaces to be rendered as the step body.
    This is less convenient, as it makes harder to work with code blocks and large(r) text bodies.

2. Step 2

    Some step2 text, indented as well.
///
```

And much like the tabs or blocks extension it would be amazing if there was an example how one can transform the approach outlined above to a more idiomatic block syntax:

```
/// step | Step 1
Some text that is not indented, and it makes it easy to write the step body.
///

/// step | Step 2
Some step 2 body
///
```

Hope this clears up the original question.

**Reply by facelessuser** [reply]
> And much like the tabs or blocks extension it would be amazing if there was an example how one can transform the approach outlined above to a more idiomatic block syntax:

Yes, what you want can be done, but it is non-trivial. What you want requires a bit of logic to create a nested structure that is aware if a previously defined structure already exists and needs to be appended to. So yes, it is possible, but it is more complicated, like the Tab extension. A solution like Tab could be adapted, but I don't have the time to write up a custom extension as an example to show this. I guess, in a sense, the Tab extension is such an example and can be referenced for those who are more adventurous.

Wrapping ordered lists in an element is a trivial approach that works out of the box and is intuitive in the sense that a steps list is indeed, still a list. I get that some people may desire a more explicit, specific object for a step that differs from list syntax. I also understand that some people find indentation annoying (I certainly did with admonitions, details, etc.), but because wrapping a list with a simple tag is so easy and gives me what I want, I don't think I would personally have interest spending more time on an alternative approach.

**Reply by hellt** [reply]
Hi @facelessuser
I have been using the approach you shared in https://github.com/squidfunk/mkdocs-material/issues/7760#issuecomment-2574390090 and it worked nicely until I upgraded the insiders from 9.5.45-insiders-4.53.14 to 9.6, which changed the pymdown extensions from 10.12 to 10.14

Then the nested tabs in the steps stopped working.

I used to have them defined as such:

````
4. **Prepare the VM/Server**

    If you are deploying the EDA Playground on a VM/Server, you should take care of the following:

    1. Install `make` that orchestrates the installation of the EDA Playground.

        /// tab | apt

        ```shell
        sudo apt install -y make
        ```

        ///
        /// tab | yum

        ```shell

````

which was nicely rendered as:

![CleanShot 2025-04-27 at 11 35 20@2x](https://github.com/user-attachments/assets/351cdd89-1b0d-434d-b3b7-5e708233e522)

After the upgrade, the nested tabs stopped working in the steps html wrapper:

![CleanShot 2025-04-27 at 11 36 43@2x](https://github.com/user-attachments/assets/9f805713-59aa-4221-a5f3-b08361e8efea)

Appreciate any pointers

**Reply by facelessuser** [reply]
@hellt If you feel there is a bug, please file a bug report over on Pymdown Extensions with a complete, reproducible example, noting your exact settings. There may very well be a bug.

**Comment by joapuiib** [comment]
@A3Bagged I like very much your steps CSS.

> Edit: I tagged the wrong person 🤦🏼‍♂️ 

I have achieved something similar in here: https://joapuiib.github.io/mkdocs-material-joapuiib/features/time-line/, but I prefer your styling. 
Would you mind sharing it?

---

In case its helpful for you, I'm using Block Extension for inserting a `<div>` with a specific class for styling.

-  Markdown: https://github.com/joapuiib/mkdocs-material-joapuiib/blob/main/docs/features/time-line.md?plain=1
- CSS: https://github.com/joapuiib/mkdocs-material-joapuiib/blob/main/material_joapuiib/templates/stylesheets/time-line.css

```
/// html | div.timeline.dashed
### Dashed line
This is a timeline item with a dashed line.
///

/// html | div.timeline.error.cross
### Error and dotted item
This is a timeline item with an error and a cross mark.
///
```
